crates/core/src/dom/mod.rs

```diff
@@ -507,8 +494,184 @@ impl Document {
         let printer = Printer::new(self, node, options);
         printer.print(writer)
     }
+
+    pub fn merge_fragment(&mut self, root_diff: RootDiff) -> Result<Root, MergeError> {
+        let root = if let Some(root) = &self.fragment_template {
+            let mut root = root.as_ref().lock().unwrap();
+            *root = root.clone().merge(root_diff)?;
+            root.clone()
+        } else {
+            let new_root : Root = root_diff.try_into()?;
+            self.fragment_template = Some(Arc::new(Mutex::new(new_root.clone())));
+            new_root
+        };
+        Ok(root)
+    }
+}
+use crate::diff::PatchResult;
+impl Document {
```

Comment:
Why two `impl Document`?

Reply:
Ah yes, I was trying to the proc-macro. I'll put it in one.